<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>scReadSim.Utility &mdash; scReadSim  documentation</title><link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  <script id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> scReadSim
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption"><span class="caption-text">Overview:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../API.html">API</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">scReadSim</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../index.html">Module code</a> &raquo;</li>
      <li>scReadSim.Utility</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for scReadSim.Utility</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">csv</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pysam</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
<span class="kn">from</span> <span class="nn">time</span> <span class="kn">import</span> <span class="n">process_time</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span> <span class="nn">os</span>

<div class="viewcode-block" id="CallPeak"><a class="viewcode-back" href="../../scReadSim.html#scReadSim.Utility.CallPeak">[docs]</a><span class="k">def</span> <span class="nf">CallPeak</span><span class="p">(</span><span class="n">macs3_directory</span><span class="p">,</span> <span class="n">INPUT_bamfile</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">MACS3_peakname_pre</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Perform peak calling using MACS3 </span>

<span class="sd">    Arguments</span>
<span class="sd">    ----------</span>
<span class="sd">    macs3_directory: `str`</span>
<span class="sd">        Directory of software MACS3.</span>
<span class="sd">    INPUT_bamfile: `str`</span>
<span class="sd">        Directory of input BAM file.</span>
<span class="sd">    outdirectory: `str`</span>
<span class="sd">        Output directory of peak calling.</span>
<span class="sd">    MACS3_peakname_pre: `str`</span>
<span class="sd">        Base name of peak calling results for MACS3.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">macs_cmd</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/macs3 callpeak -f BAMPE -t </span><span class="si">%s</span><span class="s2"> -g mm -n </span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2"> -B -q 0.01 --outdir </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">macs3_directory</span><span class="p">,</span> <span class="n">INPUT_bamfile</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">MACS3_peakname_pre</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">)</span>
    <span class="n">output</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">macs_cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">executable</span><span class="o">=</span><span class="s2">&quot;/bin/bash&quot;</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">error</span><span class="p">:</span>
         <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[ERROR] Fail to call peaks:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">error</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span></div>


<div class="viewcode-block" id="ExtractBAMCoverage"><a class="viewcode-back" href="../../scReadSim.html#scReadSim.Utility.ExtractBAMCoverage">[docs]</a><span class="k">def</span> <span class="nf">ExtractBAMCoverage</span><span class="p">(</span><span class="n">INPUT_bamfile</span><span class="p">,</span> <span class="n">samtools_directory</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Examine the covered chromosome names for the input bam file.</span>

<span class="sd">    Arguments</span>
<span class="sd">    ----------</span>
<span class="sd">    INPUT_bamfile: `str`</span>
<span class="sd">        Directory of input BAM file.</span>
<span class="sd">    samtools_directory: `str`</span>
<span class="sd">        Directory of software samtools.</span>
<span class="sd">    outdirectory: `str`</span>
<span class="sd">        Output directory.</span>
<span class="sd">    </span>
<span class="sd">    Return</span>
<span class="sd">    ------</span>
<span class="sd">    chromosomes_coverd: `list`</span>
<span class="sd">        List of chromosome names that the input bam files covers.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/samtools idxstats </span><span class="si">%s</span><span class="s2"> &gt; </span><span class="si">%s</span><span class="s2">/bam.stats.txt&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">samtools_directory</span><span class="p">,</span> <span class="n">INPUT_bamfile</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">)</span>
    <span class="n">output</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">executable</span><span class="o">=</span><span class="s2">&quot;/bin/bash&quot;</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">error</span><span class="p">:</span>
         <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[ERROR] Fail to rerturn the index BAM information:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">error</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span>    
    <span class="n">bamstats</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/bam.stats.txt&quot;</span> <span class="o">%</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">chromosomes_coverd</span> <span class="o">=</span> <span class="n">bamstats</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">bamstats</span><span class="p">[:,</span><span class="mi">2</span><span class="p">])[</span><span class="mi">0</span><span class="p">],</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">chromosomes_coverd</span></div>


<div class="viewcode-block" id="scATAC_CreateFeatureSets"><a class="viewcode-back" href="../../scReadSim.html#scReadSim.Utility.scATAC_CreateFeatureSets">[docs]</a><span class="k">def</span> <span class="nf">scATAC_CreateFeatureSets</span><span class="p">(</span><span class="n">INPUT_bamfile</span><span class="p">,</span> <span class="n">samtools_directory</span><span class="p">,</span> <span class="n">bedtools_directory</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">genome_size_file</span><span class="p">,</span> <span class="n">ref_peakfile</span><span class="p">,</span> <span class="n">ref_comple_peakfile</span><span class="p">,</span> <span class="n">MACS3_peakname_pre</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create the foreground and background feature set for the input scATAC-seq bam file.</span>

<span class="sd">    Arguments</span>
<span class="sd">    ----------</span>
<span class="sd">    INPUT_bamfile: `str`</span>
<span class="sd">        Directory of input BAM file.</span>
<span class="sd">    samtools_directory: `str`</span>
<span class="sd">        Directory of software samtools.</span>
<span class="sd">    bedtools_directory: `str`</span>
<span class="sd">        Directory of software bedtools.</span>
<span class="sd">    outdirectory: `str`</span>
<span class="sd">        Output directory.</span>
<span class="sd">    genome_size_file: `str`</span>
<span class="sd">        Directory of Genome sizes file. The file should be a tab delimited text file with two columns: first column for the chromosome name, second column indicates the size.</span>
<span class="sd">    ref_peakfile: `str`</span>
<span class="sd">        Specify the base name of output foreground feature bed file.</span>
<span class="sd">    ref_comple_peakfile: `str`</span>
<span class="sd">        Specify the base name of output background feature bed file.    </span>
<span class="sd">    MACS3_peakname_pre: `str`</span>
<span class="sd">        Base name of peak calling results for MACS3.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/bedtools sort -i </span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">_peaks.narrowPeak | </span><span class="si">%s</span><span class="s2">/bedtools merge  &gt; </span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">bedtools_directory</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">MACS3_peakname_pre</span><span class="p">,</span> <span class="n">bedtools_directory</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">ref_peakfile</span><span class="p">)</span>
    <span class="n">output</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">executable</span><span class="o">=</span><span class="s2">&quot;/bin/bash&quot;</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">error</span><span class="p">:</span>
         <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[ERROR] Fail to create feature set:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">error</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span>
    <span class="n">chromosomes_coverd</span> <span class="o">=</span> <span class="n">ExtractBAMCoverage</span><span class="p">(</span><span class="n">INPUT_bamfile</span><span class="p">,</span> <span class="n">samtools_directory</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">)</span>
    <span class="n">search_string_chr</span> <span class="o">=</span> <span class="s1">&#39;|&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">chromosomes_coverd</span><span class="p">)</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s2">&quot;cat </span><span class="si">%s</span><span class="s2"> | grep -Ew &#39;</span><span class="si">%s</span><span class="s2">&#39; &gt; </span><span class="si">%s</span><span class="s2">/genome_size_selected.txt&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">genome_size_file</span><span class="p">,</span> <span class="n">search_string_chr</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">)</span>
    <span class="n">output</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">executable</span><span class="o">=</span><span class="s2">&quot;/bin/bash&quot;</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">error</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[ERROR] Fail to extract gene regions from genome annotation file:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">error</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span>
    <span class="n">complement_cmd</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/bedtools complement -i </span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2"> -g </span><span class="si">%s</span><span class="s2">/genome_size_selected.txt &gt; </span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">bedtools_directory</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">ref_peakfile</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">ref_comple_peakfile</span><span class="p">)</span>
    <span class="n">output</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">complement_cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">executable</span><span class="o">=</span><span class="s2">&quot;/bin/bash&quot;</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">error</span><span class="p">:</span>
         <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[ERROR] Fail to create complementary feature set:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">error</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Done!</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>   </div>


<div class="viewcode-block" id="scRNA_CreateFeatureSets"><a class="viewcode-back" href="../../scReadSim.html#scReadSim.Utility.scRNA_CreateFeatureSets">[docs]</a><span class="k">def</span> <span class="nf">scRNA_CreateFeatureSets</span><span class="p">(</span><span class="n">INPUT_bamfile</span><span class="p">,</span> <span class="n">samtools_directory</span><span class="p">,</span> <span class="n">bedtools_directory</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">genome_annotation</span><span class="p">,</span> <span class="n">genome_size_file</span><span class="p">,</span> <span class="n">ref_peakfile</span><span class="p">,</span> <span class="n">ref_comple_peakfile</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create the foreground and background feature set for the input scRNA-seq bam file.</span>

<span class="sd">    Arguments</span>
<span class="sd">    ----------</span>
<span class="sd">    INPUT_bamfile: `str`</span>
<span class="sd">        Directory of input BAM file.</span>
<span class="sd">    samtools_directory: `str`</span>
<span class="sd">        Directory of software samtools.</span>
<span class="sd">    bedtools_directory: `str`</span>
<span class="sd">        Directory of software bedtools.</span>
<span class="sd">    outdirectory: `str`</span>
<span class="sd">        Output directory.</span>
<span class="sd">    genome_annotation: `str`</span>
<span class="sd">        Directory of Genome annotation file for the reference genome that the input bamfile aligned on.</span>
<span class="sd">    genome_size_file: `str`</span>
<span class="sd">        Directory of Genome sizes file. The file should be a tab delimited text file with two columns: first column for the chromosome name, second column indicates the size.</span>
<span class="sd">    ref_peakfile: `str`</span>
<span class="sd">        Specify the base name of output foreground feature bed file.</span>
<span class="sd">    ref_comple_peakfile: `str`</span>
<span class="sd">        Specify the base name of output background feature bed file.    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genome_size_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">genome_size_file</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">chromosomes_coverd</span> <span class="o">=</span> <span class="n">ExtractBAMCoverage</span><span class="p">(</span><span class="n">INPUT_bamfile</span><span class="p">,</span> <span class="n">samtools_directory</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">)</span>
    <span class="n">search_string_chr</span> <span class="o">=</span> <span class="s1">&#39;|&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">chromosomes_coverd</span><span class="p">)</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s2">&quot;cat </span><span class="si">%s</span><span class="s2"> | grep -Ew &#39;</span><span class="si">%s</span><span class="s2">&#39; &gt; </span><span class="si">%s</span><span class="s2">/genome_size_selected.txt&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">genome_size_file</span><span class="p">,</span> <span class="n">search_string_chr</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">)</span>
    <span class="n">output</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">executable</span><span class="o">=</span><span class="s2">&quot;/bin/bash&quot;</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">error</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[ERROR] Fail to extract corresponding chromosomes from genome size file:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">error</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;awk -F&quot;</span><span class="se">\t</span><span class="s2">&quot; &#39;$3==&quot;gene&quot;&#39; </span><span class="si">%s</span><span class="s2"> | cut -f1,4,5 &gt; </span><span class="si">%s</span><span class="s2">/gene_region.bed&quot;&quot;&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">genome_annotation</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">)</span>
    <span class="n">output</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">executable</span><span class="o">=</span><span class="s2">&quot;/bin/bash&quot;</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">error</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[ERROR] Fail to extract gene regions from genome annotation file:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">error</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/bedtools sort -i </span><span class="si">%s</span><span class="s2">/gene_region.bed | </span><span class="si">%s</span><span class="s2">/bedtools merge | grep -Ew &#39;</span><span class="si">%s</span><span class="s2">&#39; &gt; </span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">bedtools_directory</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">bedtools_directory</span><span class="p">,</span> <span class="n">search_string_chr</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">ref_peakfile</span><span class="p">)</span>
    <span class="n">output</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">executable</span><span class="o">=</span><span class="s2">&quot;/bin/bash&quot;</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">error</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[ERROR] Fail to create feature set:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">error</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span>
    <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s2">&quot;rm </span><span class="si">%s</span><span class="s2">/gene_region.bed&quot;</span> <span class="o">%</span> <span class="n">outdirectory</span><span class="p">)</span>
    <span class="n">complement_cmd</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/bedtools complement -i </span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2"> -g </span><span class="si">%s</span><span class="s2">/genome_size_selected.txt &gt; </span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">bedtools_directory</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">ref_peakfile</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">ref_comple_peakfile</span><span class="p">)</span>
    <span class="n">output</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">complement_cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">executable</span><span class="o">=</span><span class="s2">&quot;/bin/bash&quot;</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">error</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[ERROR] Fail to create complementary feature set:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">error</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Done!</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>   </div>


<div class="viewcode-block" id="bam2countmat"><a class="viewcode-back" href="../../scReadSim.html#scReadSim.Utility.bam2countmat">[docs]</a><span class="k">def</span> <span class="nf">bam2countmat</span><span class="p">(</span><span class="n">cells_barcode_file</span><span class="p">,</span> <span class="n">bed_file</span><span class="p">,</span> <span class="n">INPUT_bamfile</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">count_mat_filename</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Construct count matrix.</span>

<span class="sd">    Arguments</span>
<span class="sd">    ----------</span>
<span class="sd">    cells_barcode_file: `str`</span>
<span class="sd">        Directory of the cell barcode file corresponding to the input BAM file.</span>
<span class="sd">    bed_file: `str`</span>
<span class="sd">        Directory of the features bed file that the count matrix constructs based on.</span>
<span class="sd">    INPUT_bamfile: `str`</span>
<span class="sd">        Directory of input BAM file.</span>
<span class="sd">    outdirectory: `str`</span>
<span class="sd">        Output directory.</span>
<span class="sd">    count_mat_filename: `str`</span>
<span class="sd">        Specify the base name of output count matrix.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cells</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cells_barcode_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">cells</span> <span class="o">=</span> <span class="n">cells</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">cells_barcode</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">cells</span><span class="p">]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">outdirectory</span><span class="p">,</span> <span class="n">count_mat_filename</span><span class="p">),</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outsfile</span><span class="p">:</span>
        <span class="n">samfile</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">INPUT_bamfile</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">bed_directory</span><span class="p">,</span> <span class="n">bed_file</span><span class="p">))</span> <span class="k">as</span> <span class="n">open_peak</span><span class="p">:</span>
            <span class="n">reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">open_peak</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">open_peak</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">reader</span><span class="p">))</span>
        <span class="n">k</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">cellsdic</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="k">lambda</span><span class="p">:</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">cells_barcode</span><span class="p">:</span>
            <span class="n">cellsdic</span><span class="p">[</span><span class="n">cell</span><span class="p">]</span> <span class="o">=</span> <span class="n">k</span>
            <span class="n">k</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">k</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">peaksdic</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="k">lambda</span><span class="p">:</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">open_peak</span><span class="p">:</span>
            <span class="n">rec_name</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">rec</span><span class="p">)</span>
            <span class="n">peaksdic</span><span class="p">[</span><span class="n">rec_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">k</span>
            <span class="n">k</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">cells_n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cells_barcode</span><span class="p">)</span>
        <span class="n">peaks_n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">open_peak</span><span class="p">)</span>
        <span class="c1"># marginal_count_vec = [0] * len(open_peak)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Converting count matrix...</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="c1"># for rec in open_peak:</span>
        <span class="k">for</span> <span class="n">rec_id</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">open_peak</span><span class="p">))):</span>
            <span class="n">rec</span> <span class="o">=</span> <span class="n">open_peak</span><span class="p">[</span><span class="n">rec_id</span><span class="p">]</span>
            <span class="n">rec_name</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">rec</span><span class="p">)</span>
            <span class="n">currcounts</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="n">cells_n</span>
            <span class="n">reads</span> <span class="o">=</span> <span class="n">samfile</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">rec</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">rec</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">rec</span><span class="p">[</span><span class="mi">2</span><span class="p">]))</span>
            <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">:</span>
                <span class="n">cell</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">qname</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">cells_barcode</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">currcounts</span><span class="p">[</span><span class="n">cellsdic</span><span class="p">[</span><span class="n">cell</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                        <span class="k">pass</span>
            <span class="c1"># marginal_count_vec[rec_id] = sum(currcounts)</span>
            <span class="c1"># if sum(currcounts) &gt; 0:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">rec_name</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">currcounts</span><span class="p">]),</span><span class="n">file</span> <span class="o">=</span> <span class="n">outsfile</span><span class="p">)</span></div>


<div class="viewcode-block" id="bam2countmat_INPUT"><a class="viewcode-back" href="../../scReadSim.html#scReadSim.Utility.bam2countmat_INPUT">[docs]</a><span class="k">def</span> <span class="nf">bam2countmat_INPUT</span><span class="p">(</span><span class="n">cells_barcode_file</span><span class="p">,</span> <span class="n">bed_file</span><span class="p">,</span> <span class="n">INPUT_bamfile</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">count_mat_filename</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Construct count matrix for task with user input features set. </span>

<span class="sd">    Arguments</span>
<span class="sd">    ----------</span>
<span class="sd">    cells_barcode_file: `str`</span>
<span class="sd">        Directory of the cell barcode file corresponding to the input BAM file.</span>
<span class="sd">    bed_file: `str`</span>
<span class="sd">        Directory of the features bed file that the count matrix constructs based on.</span>
<span class="sd">    INPUT_bamfile: `str`</span>
<span class="sd">        Directory of input BAM file.</span>
<span class="sd">    outdirectory: `str`</span>
<span class="sd">        Output directory.</span>
<span class="sd">    count_mat_filename: `str`</span>
<span class="sd">        Specify the base name of output count matrix.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cells</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cells_barcode_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">cells</span> <span class="o">=</span> <span class="n">cells</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">cells_barcode</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">cells</span><span class="p">]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">outdirectory</span><span class="p">,</span> <span class="n">count_mat_filename</span><span class="p">),</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outsfile</span><span class="p">:</span>
        <span class="n">samfile</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">INPUT_bamfile</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">bed_directory</span><span class="p">,</span> <span class="n">bed_file</span><span class="p">))</span> <span class="k">as</span> <span class="n">open_peak</span><span class="p">:</span>
            <span class="n">reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">open_peak</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">open_peak</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">reader</span><span class="p">))</span>
        <span class="n">k</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">cellsdic</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="k">lambda</span><span class="p">:</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">cells_barcode</span><span class="p">:</span>
            <span class="n">cellsdic</span><span class="p">[</span><span class="n">cell</span><span class="p">]</span> <span class="o">=</span> <span class="n">k</span>
            <span class="n">k</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">k</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">peaksdic</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="k">lambda</span><span class="p">:</span> <span class="p">[</span><span class="kc">None</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">open_peak</span><span class="p">:</span>
            <span class="n">rec_name</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">rec</span><span class="p">)</span>
            <span class="n">peaksdic</span><span class="p">[</span><span class="n">rec_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">k</span>
            <span class="n">k</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">cells_n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cells_barcode</span><span class="p">)</span>
        <span class="n">peaks_n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">open_peak</span><span class="p">)</span>
        <span class="c1"># marginal_count_vec = [0] * len(open_peak)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Converting count matrix...</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="c1"># for rec in open_peak:</span>
        <span class="k">for</span> <span class="n">rec_id</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">open_peak</span><span class="p">))):</span>
            <span class="n">rec</span> <span class="o">=</span> <span class="n">open_peak</span><span class="p">[</span><span class="n">rec_id</span><span class="p">]</span>
            <span class="n">rec_name</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">rec</span><span class="p">)</span>
            <span class="n">currcounts</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="n">cells_n</span>
            <span class="n">reads</span> <span class="o">=</span> <span class="n">samfile</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">rec</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">rec</span><span class="p">[</span><span class="mi">4</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">rec</span><span class="p">[</span><span class="mi">5</span><span class="p">]))</span>
            <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">reads</span><span class="p">:</span>
                <span class="n">cell</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">qname</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">cells_barcode</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">currcounts</span><span class="p">[</span><span class="n">cellsdic</span><span class="p">[</span><span class="n">cell</span><span class="p">]]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                        <span class="k">pass</span>
            <span class="c1"># marginal_count_vec[rec_id] = sum(currcounts)</span>
            <span class="c1"># if sum(currcounts) &gt; 0:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">rec_name</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">currcounts</span><span class="p">]),</span><span class="n">file</span> <span class="o">=</span> <span class="n">outsfile</span><span class="p">)</span></div>


<div class="viewcode-block" id="find_nearest"><a class="viewcode-back" href="../../scReadSim.html#scReadSim.Utility.find_nearest">[docs]</a><span class="k">def</span> <span class="nf">find_nearest</span><span class="p">(</span><span class="n">array</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Find the index of peak from `array` with a length closest to `value`.</span>

<span class="sd">    Arguments</span>
<span class="sd">    ----------</span>
<span class="sd">    array: `numpy.array`</span>
<span class="sd">        Two-column array of peaks indicating the starting and ending positions.</span>
<span class="sd">    value: `int`</span>
<span class="sd">        Integer indicating the target peak length.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    idx: `int`</span>
<span class="sd">        Index of the peak with a length closest to `value`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">array</span><span class="p">)</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">array</span> <span class="o">-</span> <span class="n">value</span><span class="p">))</span><span class="o">.</span><span class="n">argmin</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">idx</span></div>


<div class="viewcode-block" id="match_peak"><a class="viewcode-back" href="../../scReadSim.html#scReadSim.Utility.match_peak">[docs]</a><span class="k">def</span> <span class="nf">match_peak</span><span class="p">(</span><span class="n">true_peakfile</span><span class="p">,</span> <span class="n">ref_peakfile</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">assignment_file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Find the reference features for the given features `true_peakfile` according to the similarity of peak length. The assignment would be stored as `assignment_file` within `outdirectory`.</span>

<span class="sd">    Arguments</span>
<span class="sd">    ----------</span>
<span class="sd">    true_peakfile: `str`</span>
<span class="sd">        Directory of the target features bed file.</span>
<span class="sd">    ref_peakfile: `str`</span>
<span class="sd">        Directory of the reference features bed file.</span>
<span class="sd">    outdirectory: `str`</span>
<span class="sd">        Output directory.</span>
<span class="sd">    assignment_file: `str`</span>
<span class="sd">        Specify the name of peak assignment file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">true_peakfile</span><span class="p">)</span> <span class="k">as</span> <span class="n">true_peak_file</span><span class="p">:</span>
        <span class="n">reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">true_peak_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">true_peak_set</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">reader</span><span class="p">))</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">ref_peakfile</span><span class="p">)</span> <span class="k">as</span> <span class="n">ref_peak_file</span><span class="p">:</span>
        <span class="n">reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">ref_peak_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">ref_peak_set</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">reader</span><span class="p">))</span>
    <span class="n">ref_peak_fraglen</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ref_peak_set</span><span class="p">])</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">outdirectory</span><span class="p">,</span> <span class="n">assignment_file</span><span class="p">),</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outsfile</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">true_peak</span> <span class="ow">in</span> <span class="n">true_peak_set</span><span class="p">:</span>
            <span class="n">true_length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">true_peak</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="n">true_peak</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">idx</span> <span class="o">=</span> <span class="n">find_nearest</span><span class="p">(</span><span class="n">ref_peak_fraglen</span><span class="p">,</span> <span class="n">true_length</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">true_peak</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ref_peak_set</span><span class="p">[</span><span class="n">idx</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]),</span> <span class="n">file</span><span class="o">=</span><span class="n">outsfile</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Done!&#39;</span><span class="p">)</span></div>


<div class="viewcode-block" id="ComplementFeature"><a class="viewcode-back" href="../../scReadSim.html#scReadSim.Utility.ComplementFeature">[docs]</a><span class="k">def</span> <span class="nf">ComplementFeature</span><span class="p">(</span><span class="n">feature_file</span><span class="p">,</span> <span class="n">comple_faeture_peakfile</span><span class="p">,</span> <span class="n">genome_size_file</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">bedtools_directory</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create background feature set given the user input foreground feature set `feature_file`.</span>

<span class="sd">    Arguments</span>
<span class="sd">    ----------</span>
<span class="sd">    feature_file: `str`</span>
<span class="sd">        Directory of the input feature set.</span>
<span class="sd">    comple_faeture_peakfile: `str`</span>
<span class="sd">        Specify the base name of output background feature bed file.    </span>
<span class="sd">    genome_size_file: `str`</span>
<span class="sd">        Directory of Genome sizes file. The file should be a tab delimited text file with two columns: first column for the chromosome name, second column indicates the size.</span>
<span class="sd">    outdirectory: `str`</span>
<span class="sd">        Output directory.</span>
<span class="sd">    bedtools_directory: `str`</span>
<span class="sd">        Directory of software bedtools.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genome_size_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">genome_size_file</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">input_peak_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">feature_file</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">chromosomes_coverd</span> <span class="o">=</span> <span class="n">input_peak_df</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="c1"># Select chromosize according to input bed file</span>
    <span class="n">search_string_chr</span> <span class="o">=</span> <span class="s1">&#39;|&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">chromosomes_coverd</span><span class="p">)</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s2">&quot;cat </span><span class="si">%s</span><span class="s2"> | grep -Ew &#39;</span><span class="si">%s</span><span class="s2">&#39; &gt; </span><span class="si">%s</span><span class="s2">/genome_size_selected.txt&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">genome_size_file</span><span class="p">,</span> <span class="n">search_string_chr</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">)</span>
    <span class="n">output</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">executable</span><span class="o">=</span><span class="s2">&quot;/bin/bash&quot;</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">error</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[ERROR] Fail to extract corresponding chromosomes from genome size file:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">error</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span>
    <span class="n">complement_cmd</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">/bedtools complement -i </span><span class="si">%s</span><span class="s2"> -g </span><span class="si">%s</span><span class="s2">/genome_size_selected.txt &gt; </span><span class="si">%s</span><span class="s2">/</span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">bedtools_directory</span><span class="p">,</span> <span class="n">feature_file</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">outdirectory</span><span class="p">,</span> <span class="n">comple_faeture_peakfile</span><span class="p">)</span>
    <span class="n">output</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">complement_cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">executable</span><span class="o">=</span><span class="s2">&quot;/bin/bash&quot;</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">error</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;[ERROR] Fail to create complementary feature set:</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">error</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Done!&#39;</span><span class="p">)</span>   </div>
















</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Guanao Yan.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>